### ISSUE TITLE
Enums with only one variant and without #[repr(_)] have size 0 in 1.27 beta

### ISSUE THREAD

**Issue by pietroalbini** [rust-lang/rust#50830]
Some crates are defining enums with only one variant without using `#[repr(_)]`, and in 1.27 beta those enums have size 0. This breaks those crates when they use transmute though.

An easy fix for those crates is adding the appropriate `#[repr()]` attribute to the enums, but we might want to revert the change to avoid breaking compatibility.

```
// Taken from libvirt-rpc in src/request.rs

#[derive(Debug)]
pub enum EventShutdownDetailType 
    Finished = 0,
}
```

* [libvirt-rpc](https://crates.io/crates/libvirt-rpc) regressed from stable to beta ([build log](https://cargobomb-reports.s3.amazonaws.com/beta-1.27-1/beta-2018-05-10/reg/libvirt-rpc-0.1.11/log.txt)) cc @polachok
* [utp](https://crates.io/crates/utp) regressed from stable to beta ([build log](https://cargobomb-reports.s3.amazonaws.com/beta-1.27-1/beta-2018-05-10/reg/utp-0.6.3/log.txt)) cc @meqif
* [temp_utp](https://crates.io/crates/temp_utp) regressed from stable to beta ([build log](https://cargobomb-reports.s3.amazonaws.com/beta-1.27-1/beta-2018-05-10/reg/temp_utp-0.8.1/log.txt)) cc @vinipsmaker

**Comment by rkruppe** [comment]
cc @nox

**Comment by nox** [comment]
That was expected, I just don't think those `transmute` should be done in the first place. There is no guarantee that the enum has the size you expect it to have.

**Comment by eddyb** [comment]
One such unsound transmute was the cause for one of the regressions from #45225, where an integer was transmuted to an `enum` even when it didn't match any of the `enum` variants, and that resulted in a crash *elsewhere* because the value was used as a niche.
I agree with @nox that catching these *is good*, but @rust-lang/compiler might have something to say.

**Comment by eddyb** [comment]
Also, all such problem can be fixed "locally", by adding a `match` or `if-else` chain and panicking if an incorrect value is given, without even changing the definition of the `enum`, *at all*.
So my suggestion would be to take each `x.*` and `0.y.*` group of releases that contains the bad transmutes, yank them, and release one patch version for each, without the transmutes.

**Comment by meqif** [comment]
I fully agree with the prevailing viewpoint that transmuting was undefined behavior, and that it really shouldn't be used.

In fact, this issue came up before and I fixed it in a later version of utp — I'm not sure why an older version of the crate was targeted here.

**Comment by pietroalbini** [comment]
> In fact, this issue came up before and I fixed it in a later version of utp — I'm not sure why an older version of the crate was targeted here.

Crater tests always the latest version of every crate, and also an older version for some of them.

**Comment by scottmcm** [comment]
Note that previous changes have also "broken" UB transmutes, such as `transmute::<[u8; 6], (u8,u16,u8)>`, so this is a previously-decided category of allowed change.
